File: skills/daftai-local-video-subtitler/scripts/convert_vtt_to_srt.py
#!/usr/bin/env python3
"""
将 VTT 字幕转换为 SRT 格式
"""

import sys
import re
from pathlib import Path


def vtt_to_srt(vtt_path: str, srt_path: str) -> str:
    """
    将 VTT 字幕转换为 SRT 格式
    """
    vtt_path = Path(vtt_path)
    srt_path = Path(srt_path)
    
    if not vtt_path.exists():
        raise FileNotFoundError(f"VTT file not found: {vtt_path}")
    
    print(f"🔄 转换 VTT → SRT...")
    print(f"   输入: {vtt_path}")
    print(f"   输出: {srt_path}")
    
    with open(vtt_path, 'r', encoding='utf-8') as f:
        content = f.read()
    
    # 分割成块
    blocks = content.strip().split('\n\n')
    
    srt_blocks = []
    index = 1
    
    for block in blocks:
        lines = block.strip().split('\n')
        
        # 跳过 WEBVTT 头部和元数据
        if any(line.startswith(('WEBVTT', 'Kind:', 'Language:', 'X-TIMESTAMP')) for line in lines):
            continue
        
        # 查找时间戳
        timestamp_line = None
        timestamp_idx = -1
        
        for idx, line in enumerate(lines):
            if '-->' in line:
                timestamp_line = line
                timestamp_idx = idx
                break
        
        if timestamp_line is None:
            continue
        
        # 获取文本（时间戳之后的所有行）
        text_lines = []
        for line in lines[timestamp_idx + 1:]:
            # 移除 VTT 样式标签
            clean_line = re.sub(r'<[^>]+>', '', line).strip()
            if clean_line:
                text_lines.append(clean_line)
        
        if not text_lines:
            continue
        
        # 转换时间戳格式
        # VTT: 00:00:00.000 --> 00:00:03.500
        # SRT: 00:00:00,000 --> 00:00:03,500
        
        # 先移除位置信息 (如 align:start position:0%)，在替换点号之前
        srt_timestamp = re.sub(r'\s+(align|position|line|size|vertical):\S+', '', timestamp_line)
        # 再替换点号为逗号
        srt_timestamp = srt_timestamp.replace('.', ',').strip()
        
        # 构建 SRT 块
        srt_block = f"{index}\n{srt_timestamp}\n" + '\n'.join(text_lines)
        srt_blocks.append(srt_block)
        index += 1
    
    # 写入 SRT 文件
    srt_path.parent.mkdir(parents=True, exist_ok=True)
    with open(srt_path, 'w', encoding='utf-8') as f:
        f.write('\n\n'.join(srt_blocks) + '\n')
    
    print(f"✅ 转换完成，共 {index - 1} 条字幕")
    return str(srt_path)


def main():
    if len(sys.argv) < 3:
        print("用法: python convert_vtt_to_srt.py <input.vtt> <output.srt>")
        sys.exit(1)
    
    vtt_path = sys.argv[1]
    srt_path = sys.argv[2]
    
    try:
        result = vtt_to_srt(vtt_path, srt_path)
        print(f"\n✨ 完成！输出文件: {result}")
    except Exception as e:
        print(f"\n❌ 错误: {str(e)}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: skills/daftai-video-subtitler/scripts/convert_vtt_to_srt.ts
#!/usr/bin/env bun
/**
 * 将 VTT 字幕转换为 SRT 格式
 */

import * as fs from "fs";
import * as path from "path";

export function vttToSrt(vttPath: string, srtPath: string): string {
  if (!fs.existsSync(vttPath)) {
    throw new Error(`VTT file not found: ${vttPath}`);
  }

  console.log(`🔄 转换 VTT → SRT...`);
  console.log(`   输入: ${vttPath}`);
  console.log(`   输出: ${srtPath}`);

  const content = fs.readFileSync(vttPath, "utf-8");
  const blocks = content.trim().split("\n\n");

  const srtBlocks: string[] = [];
  let index = 1;

  for (const block of blocks) {
    const lines = block.trim().split("\n");

    // 跳过 WEBVTT 头部和元数据
    if (lines.some((line) => /^(WEBVTT|Kind:|Language:|X-TIMESTAMP)/.test(line))) {
      continue;
    }

    // 查找时间戳
    let timestampLine: string | null = null;
    let timestampIdx = -1;

    for (let idx = 0; idx < lines.length; idx++) {
      if (lines[idx].includes("-->")) {
        timestampLine = lines[idx];
        timestampIdx = idx;
        break;
      }
    }

    if (timestampLine === null) continue;

    // 获取文本（时间戳之后的所有行）
    const textLines: string[] = [];
    for (let i = timestampIdx + 1; i < lines.length; i++) {
      const cleanLine = lines[i].replace(/<[^>]+>/g, "").trim();
      if (cleanLine) {
        textLines.push(cleanLine);
      }
    }

    if (textLines.length === 0) continue;

    // 转换时间戳格式
    // VTT: 00:00:00.000 --> 00:00:03.500
    // SRT: 00:00:00,000 --> 00:00:03,500
    let srtTimestamp = timestampLine.replace(
      /\s+(align|position|line|size|vertical):\S+/g,
      ""
    );
    srtTimestamp = srtTimestamp.replace(/\./g, ",").trim();

    srtBlocks.push(`${index}\n${srtTimestamp}\n${textLines.join("\n")}`);
    index++;
  }

  // 写入 SRT 文件
  const srtDir = path.dirname(srtPath);
  fs.mkdirSync(srtDir, { recursive: true });
  fs.writeFileSync(srtPath, srtBlocks.join("\n\n") + "\n", "utf-8");

  console.log(`✅ 转换完成，共 ${index - 1} 条字幕`);
  return srtPath;
}

// CLI 入口
if (require.main === module) {
  if (process.argv.length < 4) {
    console.log("用法: bun convert_vtt_to_srt.ts <input.vtt> <output.srt>");
    process.exit(1);
  }

  const vttPath = process.argv[2];
  const srtPath = process.argv[3];

  try {
    const result = vttToSrt(vttPath, srtPath);
    console.log(`\n✨ 完成！输出文件: ${result}`);
  } catch (e: any) {
    console.log(`\n❌ 错误: ${e.message}`);
    process.exit(1);
  }
}
